### ISSUE TITLE
Pretty-printer should take TDNR into account

### ISSUE THREAD

**Issue by runarorama** [unisonweb/unison#2881]
The expression `0.1 + 0.2` is unambiguous because TDNR uses the expected type of `+` to figure out that what's meant is `base.Float.+`.

But when this gets pretty-printed out again, the shortest unambiguous suffix is `Float.+`,

So this expression renders as `0.1 Float.+ 0.2`, which is terribly ugly.

It would be really nice if the pretty-printer could figure out that `+` is unambiguous here.

**Comment by pchiusano** [comment]
Well, it is by design that it doesn't do this - all symbols are unambiguous and don't rely on... talking to the typechecker during pretty-printing. I think it's a major can of worms to try to do anything else.

That said! I think when rendering to HTML, we could elide all `use` statements and print all infix operators out unqualified. The reference to the operator will have the hash in there, so its full name could be shown in a tooltip or what have you.

~~Another hack is to just special case numeric operators in the pretty-printer - if calling `Nat.+` or other builtins and both arguments are literals, don't bother to emit the qualified name and/or `use` statement.~~ Disregard, this is a bad idea and not correct in all cases - for instance, you could have a local variable called `(+)` in scope, and this would cause that to be captured. You could also have a definition called `Blah.+` in the same file, and depending on the surrounding context of the expression this could lead to a TDNR ambiguity.

**Comment by ceedubs** [comment]
> I think it's a major can of worms to try to do anything else.

@pchiusano any specific concerns here? Do you think that the implementation would become difficult to manage? One concern that I would have is performance. Parsing performance isn't all that critical since textual source code is ephemeral and isn't parsed often. But pretty-printing probably happens more often (certainly on Share it happens a lot).

 > when rendering to HTML, we could elide all use statements and print all infix operators out unqualified
 
 I can see how this might be nice for displaying clean code samples. But I love that currently I can copy/paste code from Share/blog posts and be pretty confident that it will compile exactly as-is for me locally. I think that we should strive to maintain this behavior if possible. Perhaps you could render the code with minimal names but there's a "copy to clipboard" button that copies qualified names?

**Comment by pchiusano** [comment]
> any specific concerns here? Do you think that the implementation would become difficult to manage?

Yes. Right now the pretty-printer just needs names for each hash. That's it. 

Accurately eliding qualified names based on type signatures is hard. Even if you plumb all the information needed through to the pretty-printer (you'd need basically a typechecking environment), what is the algorithm? You can't make decisions locally - you might notice that `1.0 Float.+ x` still typechecks the same as `1.0 + x`, but then when you make another change to dequalify a symbol in the same function, suddenly your program doesn't typecheck anymore because it's ambiguous. 

> But I love that currently I can copy/paste code from Share/blog posts and be pretty confident that it will compile exactly as-is for me locally. I think that we should strive to maintain this behavior if possible. Perhaps you could render the code with minimal names but there's a "copy to clipboard" button that copies qualified names?

I like this idea.

**Comment by ceedubs** [comment]
> but then when you make another change to dequalify a symbol in the same function, suddenly your program doesn't typecheck anymore

That's a really good point. I hadn't considered that.

**Comment by mitchellwrosen** [comment]
Is this a #wontfix?

**Comment by aryairani** [comment]
I think it's a wontfix until we have a richer editor; or if we want to special-case something for Share specifically

(not sure if this ticket is about Share or about `edit`? @runarorama)
